### ISSUE TITLE
Executor hang with tokio oneshot

### ISSUE THREAD

**Issue by waynexia** [DataDog/glommio#325]
I want to use the `oneshot` channel as task's `JoinHandle`, which can be `await` on other runtimes like tokio. 

<details><summary>Here is a reproducible example</summary>

```rust

#[tokio::test]
async fn tokio_glommio() {
    use glommio::{Local, LocalExecutor};
    use tokio::sync::oneshot::{channel as oneshot, Sender as Callback};

    let (tx, rx) = channel::<Callback<()>>();
    let handle = std::thread::spawn(|| {
        let ex = LocalExecutor::default();
        ex.run(async move {
            while let Ok(o_tx) = rx.recv() {
                println!("received");
                Local::local(async {
                    println!("polled");
                    o_tx.send(()).unwrap();
                })
                // Place-1
                .detach();
            }
        })
    });

    let (o_tx, o_rx) = oneshot();
    tx.send(o_tx).unwrap();
    // Place-2
    o_rx.await.unwrap();

    drop(tx);
    handle.join().unwrap();
}

```

</details>

In this example, I can see "received" but no more other things and the program keep running.

If add a `.await` after `detach()` on Place-1 or remove the line on Place-2 the "polled" will be printed. But these aren't the solution I think.

I tried to set a breakpoint on `glommio::task::JoinHandle::drop()` but it's never hit *in my environment*.

**Comment by glommer** [comment]
thanks for trying this! I will take a look soon

**Comment by glommer** [comment]
Hello

I am not able to run this:
```
 |
5 |     let (tx, rx) = channel::<Callback<()>>();
  |                    ^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
1 | use std::sync::mpsc::channel;
  |
1 | use tokio::sync::broadcast::channel;
  |
1 | use tokio::sync::mpsc::channel;
  |
1 | use tokio::sync::oneshot::channel;
  |
    and 1 other candidate
 ```

If I change `channel` to `oneshot`, then I get a different message:

```
  |
9 |             while let Ok(o_tx) = rx.recv() {
  |                                     ^^^^ method not found in `tokio::sync::oneshot::Receiver<tokio::sync::oneshot::Sender<()>>`

error: aborting due to previous error
```

I suspect this could be due to differences in tokio versions.

The easiest way would be for you to send me a branch that has that code inside the `examples` directory, with the exact tokio version you want, etc.

**Comment by waynexia** [comment]
Sorry about that, I missed this dependence. `channel` is `std::sync::mpsc::channel`.

The example code is in https://github.com/waynexia/glommio/blob/issue325-example/examples/issue325.rs.

**Comment by glommer** [comment]
Ok, I can reproduce this locally now. Thanks!

**Comment by glommer** [comment]
Ok, I think this is because you are using the sync version of `mpsc`.

Sync primitives can't work in an async thread-per-core environment. To understand why, keep in mind that there is only one thread. If it blocks, who is to wake it?

Here's a sequence of events that leads to this hang:
* `recv()` is a *blocking* call: so when it is called, the entire thread stops until an element is received.
* the first time is "works" fine (in a real application, your performance would have tanked at this point), but once the message is received in the channel, the channels is unblocked.
* Once the channel is unblocked, you print the `received message, and *creates* a new task (note that tasks are not run as soon as they are created: they are scheduled to run later)
* If you want to see this even more clearly, add a `println` statement right after you detach the task. You will see that it prints!
* What happens next depends on the implementation and timing. If you are lucky enough, the main task will yield at this point and you will run your detached task. But that's usually not happens: your main task will still have enough quota, and will keep running. Then you reach the `recv()` statement again. The thread blocks, and never ever wakes up again.

Again, the entire *executor thread* is blocked, so it cannot run your detached task!

The tokio mpsc channel would be a better fit for this, since it will not block the thread like this.

**Comment by waynexia** [comment]
Got it, tokio mpsc works for me now.

Appreciate your detailed and clear explanation!
